### PR TITLE
fix: all data mapped to one col when zooming in

### LINKED ISSUE #16746: [Bug] Data plotted at false position after zooming with filterMode "none"
### Version

5.3.1

### Link to Minimal Reproduction

https://codesandbox.io/s/echarts-datazoom-bug-q196p9

### Steps to Reproduce

1. I have created a chart that contains both line and scatter series objects.
2. The x-axis is of **type "category"**.
3. The **filterMode** of the dataZoom feature is set to **"none"** in order to prevent the line from disappearing if only one point of the line is in view after zooming.
4. Zoom in with the dataZoom feature in such a way that only one tick of the x-axis but the entire range of the y-axis is in view.

Following picture shows the chart before zooming in:
image

### Current Behavior

After zooming in as described in step 4, the data of the other ticks is plotted at the same x-position (see picture below). 
image

### Expected Behavior

After zooming in as described in step 4, only the data in view should be plotted at the correct position (as when filterMode is not set to "none").
image

### Environment

```markdown
- OS: Windows 10
- Browser: Google Chrome 99.0.4844.51
- Framework: React 17.0.2
```

### Any additional comments?

_No response_

### LINKED ISSUE #16852: [Bug] Overlapping bars when zooming
### Version

5.3.2

### Link to Minimal Reproduction

https://echarts.apache.org/examples/en/editor.html?c=line-simple&version=5.3.2

### Steps to Reproduce

1. Use this option: 
```js
option = {
  dataZoom: [
    { filterMode: 'none', type: 'inside', xAxisIndex: [0] },
    { filterMode: 'none', type: 'inside', yAxisIndex: [0] }
  ],
  xAxis: { type: 'category', data: ['Mon', 'Tue'] },
  yAxis: { type: 'value' },
  series: [
    {
      data: [
        { value: 20, itemStyle: { color: 'green' } },
        { value: 200, itemStyle: { color: 'red' } }
      ],
      type: 'bar'
    }
  ]
};
```
2. Zoom in while mouse is over the smaller bar (green) to a zoom level where y axis has range about [5, 43].

### Current Behavior

Bars are overlapping each other.
bar-overlapping-issue

### Expected Behavior

Bars are not overlapping each other. In this example, only green bar is visible in the chart area.

### Environment

```markdown
- OS: Windows 10.
- Browser: Version 100.0.4896.75 (Official Build) (64-bit)
- Framework:
```

### Any additional comments?

_No response_

### LINKED ISSUE #17014: [Bug] dataZoom显示错误的问题。dataZoom show error
### Version

5.3.2

### Link to Minimal Reproduction

_No response_

### Steps to Reproduce

1. 使用`dataZoom`，设置`dataZoom.filterMode:'none'`
2. 拖动dataZoom手柄，缩小范围至1个数据
```js
option = {
  tooltip: {
    trigger: 'axis'
  },
  xAxis: {
    data: ['2014-07-14', '2014-07-15', '2014-07-16', '2014-07-17', '2014-07-18']
  },
  yAxis: {},
  dataZoom: [{ filterMode: 'none' }],
  series: {
    name: 'Beijing AQI',
    type: 'bar',
    data: [0, 156, 140, 133, 186]
  }
};

```

### Current Behavior

当前数据展示为所有数据中的最大值

### Expected Behavior

显示正确的数据，像`dataZoom.filterMode:'empty'`一样

### Environment

```markdown
- OS:Windows 10
- Browser:Chrome 101.0.4951.64(64bit)
- Framework:none
```

### Any additional comments?

_No response_

### PR DESCRIPTION
**Pull request by jiawulin001**
## Brief Information

This pull request is in the type of:

- [x] bug fixing
- [ ] new feature
- [ ] others

### What does this PR do?

Fix the bug that when zooming in to only one column, all data is mapped on the very data.

### Fixed issues

- fix #16746 
- fix #16852
- fix #17014 

### Before: What was the problem?

The problem usually occurs when zooming in to one category column, then all data is mapped to the very same column. Why the problem occurs can be referred to here: https://github.com/apache/echarts/issues/16746#issuecomment-1100036918

#### #16746 
| Before Zoom In | After Zoom In |
| --- | --- |
|before#16746|before-zoomIn#16746|

#### #16852 
| Before Zoom In | After Zoom In |
| --- | --- |
|before2#16746|before2-zoomIn#16746|

### After: How is it fixed in this PR?

The problem is solved by adding a condition statement in both [places](https://github.com/apache/echarts/issues/16746#issuecomment-1100036918) where bugs exist. Now Echarts will correcty treat the location of item when extent[0] = extent[1].

#### #16746 
| Before Zoom In | After Zoom In |
| --- | --- |
|before#16746|after-zoomIn#16746|

#### #16852 
| Before Zoom In | After Zoom In |
| --- | --- |
|before2#16746|after2-zoomIn#16746|

## Misc

- [ ] The API has been changed (apache/echarts-doc#xxx).
- [ ] This PR depends on ZRender changes (ecomfe/zrender#xxx).

### Related test cases or examples to use the new APIs

Attached

### Merging options

- [x] Please squash the commits into a single one when merging.

### CODE REVIEW

**Comment by pissang** [comment]
The result looks good. But could you explain the strategy here?

File: src/scale/helper.ts [hunk 1]
```diff
@@ -128,7 +128,12 @@ export function contain(val: number, extent: [number, number]): boolean {
 
 export function normalize(val: number, extent: [number, number]): number {
     if (extent[1] === extent[0]) {
-        return 0.5;
+        if (val !== extent[0]) {
```
[the remark this answers is not in the document]
Comment (jiawulin001) [reply src/scale/helper.ts:131]: My apologies I didn't explain enough in the PR and comment. Let me show you the dataflow and then the strategy would be clear.
First, we are zooming in to 'b' category and only 'b' at this chart:
before#16746
Then Echarts tries to determine the location of 'a', 'b' and 'c' column in the chart **AFTER ZOOMING**. So the `extent` here is `[1, 1]`:
https://github.com/apache/echarts/blob/a5257e38ad7776c766879c83854b1faa2771a454/src/coord/Axis.ts#L122-L133
At line 125, we enter function `normalize` which returns the **relative location** (0~1) of the column according to their location **BEFORE ZOOMING** (which is 0, 1, 2 for 'a', 'b', 'c' respectively):
https://github.com/apache/echarts/blob/a5257e38ad7776c766879c83854b1faa2771a454/src/scale/helper.ts#L129-L132
At this moment, remember that `extent` is `[1, 1]`, so no matter what `data` is, function `normalize` returns 0.5, which means all the three columns are located at the middle of the chart after zooming. 
To solve this, I add a judgement: `if (val !== extent[0])` which checks whether the column is the one we are zooming in at. If it's not, then we return a correct relative location of the column. Here for 'a', 'b', 'c', they are 0, 1, 2 before zooming and then -0.5, 0.5, 1.5 (locate at -50%, 50%, 150%) after zooming. 
before#16746

Also, another bug is preventing me to distinguish the three columns. At line 132 we enters function `linearMap`, `subRange` is 0 because r0 = r1 = 320, which result in `return (val - d0) / subDomain * subRange + r0` equaling r0 no matter what `val` is. That means all three columns still locate at 320 (0.5 * 640).
https://github.com/apache/echarts/blob/a5257e38ad7776c766879c83854b1faa2771a454/src/util/number.ts#L47-L100
To solve this, I add a check whether `subRange` is 0 and if yes then just return `val * 2 * r0`. When `subRange` is 0, `r0` is exactly half length of the chart, 0.5*640 = 320. So `val * 2 * r0` = val * chartLength would be the absolute location of corresponding relative location.

**Comment by jiawulin001** [comment]
My apologies I didn't explain enough in the PR and comment. Let me show you the dataflow and then the strategy would be clear.
First, we are zooming in to 'b' category and only 'b' at this chart:
before#16746
Then Echarts tries to determine the location of 'a', 'b' and 'c' column in the chart **AFTER ZOOMING**. So the `extent` here is `[1, 1]`:
https://github.com/apache/echarts/blob/a5257e38ad7776c766879c83854b1faa2771a454/src/coord/Axis.ts#L122-L133
At line 125, we enter function `normalize` which returns the **relative location** (0~1) of the column according to their location **BEFORE ZOOMING** (which is 0, 1, 2 for 'a', 'b', 'c' respectively):
https://github.com/apache/echarts/blob/a5257e38ad7776c766879c83854b1faa2771a454/src/scale/helper.ts#L129-L132
At this moment, remember that `extent` is `[1, 1]`, so no matter what `data` is, function `normalize` returns 0.5, which means all the three columns are located at the middle of the chart after zooming. 
To solve this, I add a judgement: `if (val !== extent[0])` which checks whether the column is the one we are zooming in at. If it's not, then we return a correct relative location of the column. Here for 'a', 'b', 'c', they are 0, 1, 2 before zooming and then -0.5, 0.5, 1.5 (locate at -50%, 50%, 150%) after zooming. 
before#16746

Also, another bug is preventing me to distinguish the three columns. At line 132 we enters function `linearMap`, `subRange` is 0 because r0 = r1 = 320, which result in `return (val - d0) / subDomain * subRange + r0` equaling r0 no matter what `val` is. That means all three columns still locate at 320 (0.5 * 640).
https://github.com/apache/echarts/blob/a5257e38ad7776c766879c83854b1faa2771a454/src/util/number.ts#L47-L100
To solve this, I add a check whether `subRange` is 0 and if yes then just return `val * 2 * r0`. When `subRange` is 0, `r0` is exactly half length of the chart, 0.5*640 = 320. So `val * 2 * r0` = val * chartLength would be the absolute location of corresponding relative location.

**Comment by pissang** [comment]
Thanks for the explaination.

> To solve this, I change the return into val - extent[0] + 0.5. Here for 'a', 'b', 'c', they are 0, 1, 2 before zooming and then -0.5, 0.5, 1.5 (locate at -50%, 50%, 150%) after zooming.

It is a good point. The problem here is we still return a in range value 0.5 even when value exceeding the extent. A more expected behavior is return < 0 when val is smaller than extent[0] and return > 1 when val is larger than extent[1] to ensure the normalized value also exceed the range [0, 1].  `val - extent[0] + 0.5` can do it in the ordinal scale, which val is integers. But when it comes with decimals, it fails. The result may still in the range of [0, 1]. So I think `val < extent[0] ? -0.5 : val > extent[0] ? 1.5 : 0.5` will solve it more generally

> To solve this, I add a check whether subRange is 0 and if yes then just return (val - d0) / subDomain * 2 * r0. When subRange is 0, r0 is exactly half length of the chart, 0.5*640 = 320 (subDomain = 1 here). So (val - d0) / subDomain * 2 * r0 = val * chartLength would be the absolute location of corresponding relative location.

Seems the behavior of very underlying method is changed to match the upper logic. It's not a proper way to fix the issue and may affect other codes that rely on this method.

**Comment by jiawulin001** [comment]
>So I think `val < extent[0] ? -0.5 : val > extent[0] ? 1.5 : 0.5` will solve it more generally

I agree with this, will update the PR.
>Seems the behavior of very underlying method is changed to match the upper logic.

For this part, I do want to discuss with you some of my thoughts.
First, the function `linearMap` is to map `val` in `domain` to a specified `range`. When `val` is in the specified `domain` and `range` is only a point, then yes `val` should all be mapped to the only value in `range`. But if `val` is not in the `domain`, logically it should not be mapped to anywhere in `range`. So as I added the condition `if ((val < Math.min(d0, d1)) || (val > Math.max(d0, d1)))`, `val` is promised not to be in `domain` and it makes sense to return a value outside the `range`, which is `val * chartWidth` here. 
Basically, I think this very underlying method has this tiny logic problem and it directly causes the bug here. If we are to use this same function in the same place, I don't think the bug can be fixed.
However, if you insist that this function is better not be touched. Alternatives are that we use another function here to fix the problem. But that would add redundancy to code since this function is almost the same as `linearMap` and only used here.

**Comment by pissang** [comment]
I didn't mean keeping the function the same. Instead it should be fixed, but in a more general way. For example, what's the behavior when I mapping values to alpha? Will the fomular

**Comment by jiawulin001** [comment]
> Instead it should be fixed, but in a general way.

I agree with you. I just scanned through the code and find `linearMap` used in all charts. So returning `(val - d0) / subDomain * 2 * r0` is indeed inproper. 
Logically, a 'val' that is not in 'domain' shall not be mapped to anywhere in 'range'. So one way to do this I think is to return `undefined` if `val` not in `domain`. This is logically correct but could possibly result in fatal errors and make echarts halt.
Another way is to deal with this special situation before `linearMap` is run. That is to check if extent is one point and `val` is out of `domain`, then just return the wanted result without running `linearMap`.
https://github.com/apache/echarts/blob/a5257e38ad7776c766879c83854b1faa2771a454/src/coord/Axis.ts#L122-L133

Which way do you think is better?

**Comment by pissang** [comment]
> Another way is to deal with this special situation before linearMap is run. That is to check if extent is one point and val is out of domain, then just return the wanted result without running linearMap.

I will prefer this way to minimize the influence.

**Comment by jiawulin001** [comment]
So I've just updated the code and now the condition that 
1. `extent` is only one point;
2. `data` is out of domain
will only be checked at axis.ts/dataToCoord so it will not affect function `linearMap`.
I still think function `linearMap` has logical problem but I am not sure how to return a value out of `range` causing no risk.

File: src/coord/Axis.ts [hunk 2]
```diff
@@ -129,7 +129,12 @@ class Axis {
             fixExtentWithBands(extent, (scale as OrdinalScale).count());
         }
 
-        return linearMap(data, NORMALIZED_EXTENT, extent, clamp);
+        if ((extent[0] === extent[1]) && (data < 0 || data > 1)) {
+            return data * extent[0] * 2;
```
Comment (pissang) [inline src/coord/Axis.ts:133]: I still think the equation used here is too tricky. It heavily depends on what the data is and can't be explained when data is not -0.5 or 1.5.

Comment (jiawulin001) [reply src/coord/Axis.ts:133]: data here is promised to be -0.5, 0.5 or 1.5 in scale.normalize() when extent[0] === extent[1]

Comment (jiawulin001) [reply src/coord/Axis.ts:133]: @pissang Hi a new issue https://github.com/apache/echarts/issues/17014 opened 2 days ago is also relavant. I think such issues can occur quite often during use of Echarts. We shall solve this problem through this PR soon.

Comment (pissang) [reply src/coord/Axis.ts:133]: Yeah this bug should be fixed. But as I mentioned in the above comment. The fomular used here is just too magical and I can't be convinced it is meaningful on all cases.  For example: 
+ Why choose mutpliying 2, not 3 or any other numbers. 
+ Here we check data < 0 || data > 0 and assume `normalize` will always return -0.5, 1.5.  But what if we tweaked the result of `normalize` and return other values instead? Will we forget to tweak the fomular to fit the expect result? It's a kind of leaky abstraction

Comment (pissang) [reply src/coord/Axis.ts:133]: If I'm convinced. Then comments should be added to explain why this fomular will work

Comment (jiawulin001) [reply src/coord/Axis.ts]: After consider over this issue, I think the main problem is still in `linearMap`. I changed it and let it return NaN under permission when `val` has exceeded `domain`. And for now the permission is by default `false` and I let it be true at Axis.ts only. After testing in the future we can open permissions at other places.
Also please check the CI test because it keeps failing at collecting changed files.
